### PR TITLE
Function and task extractor

### LINKED ISSUE #418: Kythe: extract Non-ANSI style ports in module declaration
Extend Kythe facts extractor to extract Non-ANSI ports in module declaration.

```
module foo(a, b);
    input a;
    input b;
endmodule
```

Depends on: #388

### PR DESCRIPTION
**Pull request by MinaToma**
Kythe facts extractor for function and tasks.

Coverd:
- function and task name.
- function and task call (without ports).
- function and task ports in declaration.

WIP #390 
Fixes #418

### CODE REVIEW

File: verilog/CST/functions.h
Comment (fangism) [inline verilog/CST/functions.h:124]: BlockStatementList (spelling)

File: verilog/CST/functions.h
Comment (fangism) [inline verilog/CST/functions.h:118]: token

File: verilog/CST/functions.h
Comment (fangism) [inline verilog/CST/functions.h:120]: How about `GetFunctionCallNameTokenInfo`?

File: verilog/CST/tasks.h
Comment (fangism) [inline verilog/CST/tasks.h:43]: StatementList (spelling)

File: verilog/tools/kythe/indexing_facts_tree_extractor.h
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.h:65]: Suggest naming this `ExtractFunctionDeclaration`

File: verilog/tools/kythe/indexing_facts_tree_extractor.h
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.h:69]: Suggest naming this `ExtractFunctionDeclaration`

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:336]: `const ...&`

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:342]: Is there a different fact type for variables that come through ports or parameters?

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:361]: function

File: verilog/tools/kythe/indexing_facts_tree_extractor.h
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.h:75]: should comment reference "ports"?

File: verilog/tools/kythe/kythe_facts_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.cc:64]: `ExtractFunctionOrTaskCall`? (spelling)

File: verilog/tools/kythe/kythe_facts_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.cc:230]: `ExtractFunctionOrTaskCall`

File: verilog/tools/kythe/kythe_facts_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.cc:225]: Will these change to your new `Generate` functions once those are merged?

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (MinaToma) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:342]: Yes, in Kythe function can have `param` facts.
For now I am not using `param` facts so I decided to treat the parameters as variables.

Comment (MinaToma) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:342]: The `param` fact is an extra fact it won't remove the variable fact.

File: verilog/tools/kythe/kythe_facts_extractor.cc
Comment (MinaToma) [reply verilog/tools/kythe/kythe_facts_extractor.cc:225]: Yes.

File: verilog/CST/functions.cc
Comment (ivan444) [inline verilog/CST/functions.cc:83]: this should be auto*

File: verilog/CST/tasks.cc
Comment (ivan444) [inline verilog/CST/tasks.cc:52]: this should be auto*

**Review by ivan444 — APPROVED** [review]

File: verilog/CST/functions_test.cc
Comment (fangism) [inline verilog/CST/functions_test.cc:401]: Suggested test cases:

* negative test (non-functions)
* functions with return type specified
* functions with (parameters...)
* function inside module
* function inside package
* function inside class

File: verilog/CST/port.cc
Comment (fangism) [inline verilog/CST/port.cc:52]: Unit tests for this function.

File: verilog/CST/tasks_test.cc
Comment (fangism) [inline verilog/CST/tasks_test.cc:230]: Suggested test cases:

* negative (non-task)
* tasks with ports
* task inside module
* task inside package
* task inside class

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:194]: // summarize with a comment that shows the difference between ANSI (with direction/type) and non-ANSI style port.

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (fangism) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:356]: Down the road, I do think the distinction between port/parameter and local variable would be helpful.  Can you leave a //TODO?

File: verilog/tools/kythe/indexing_facts_tree_extractor_test.cc
Comment (fangism) [inline verilog/tools/kythe/indexing_facts_tree_extractor_test.cc:421]: suggest comment:
// Normally, tools will reject non-ANSI port declarations that are missing their full definitions inside the body like "input a", but here we don't care and are just checking for references, even if they are dangling.

File: verilog/tools/kythe/kythe_schema_constants.h
Comment (fangism) [inline verilog/tools/kythe/kythe_schema_constants.h:34]: Could we add

```
constexpr absl::string_view kNodeTask = "task";
```

File: verilog/tools/kythe/kythe_facts_extractor.h
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.h:137]: Would it make more sense to separate these into `ExtractFunctionDefinition` and `ExtractTaskDefinition`?  Even if they are similar today?

File: verilog/tools/kythe/kythe_facts_extractor.h
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.h:174]: Same question, can we separate their signatures?
For instance, functions have return types, and tasks do not.

**Review by fangism — CHANGES_REQUESTED** [review]
Looking good.

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc
Comment (MinaToma) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:356]: This will most likely change after merging the PRs but I left a TODO for now.

File: verilog/tools/kythe/kythe_schema_constants.h
Comment (MinaToma) [reply verilog/tools/kythe/kythe_schema_constants.h:34]: No Kythe doesn't have something called tasks.
We are treating tasks as functions in Kythe.

File: verilog/tools/kythe/kythe_facts_extractor.h
Comment (MinaToma) [reply verilog/tools/kythe/kythe_facts_extractor.h:137]: I don't think so because Kythe has nothing called `task` so the best thing is to insure that tasks are treated as functions because this is the best representation for it in kythe.

File: verilog/tools/kythe/kythe_facts_extractor.h
Comment (MinaToma) [reply verilog/tools/kythe/kythe_facts_extractor.h:174]: We don't give function return type in Kythe because we aren't interested in it at this time.

File: verilog/CST/functions_test.cc
Comment (fangism) [inline verilog/CST/functions_test.cc:418]: Btw, you can always check https://github.com/google/verible/blob/master/verilog/parser/verilog_parser_unittest.cc for more ideas.

Thought of some more that are common:

* `virtual` functions inside classes, include `pure virtual`
* `static` functions inside classes (you've covered `automatic` already)
* `import "DPI-C"` functions (see LRM 35.5.4 Import declarations). Or file a TODO for these for later because they may require a little more work.

File: verilog/CST/port_test.cc
Comment (fangism) [inline verilog/CST/port_test.cc:271]: For future work (new issue, if there isn't one already):

Facts on this example should include that `b` has the same direction and type as `a`: `input` and `wire`.

File: verilog/CST/tasks_test.cc
Comment (fangism) [inline verilog/CST/tasks_test.cc:245]: Also:

* `virtual` task inside class)
* `static` task (inside and outside class)

**Review by fangism — CHANGES_REQUESTED** [review]
In great shape!

File: verilog/tools/kythe/kythe_schema_constants.h
Comment (fangism) [reply verilog/tools/kythe/kythe_schema_constants.h:34]: Ok, we'll make distinguishing between functions/tasks a future issue.

File: verilog/CST/tasks_test.cc
Comment (MinaToma) [reply verilog/CST/tasks_test.cc:245]: @fangism  Is it possible to write a static task or function outside classes ?
I tried that but got syntax errors.

**Comment by MinaToma** [comment]
For `pure virtual` I wrote that in my TODO and will put an issue for it after this merge because I think it's related to class hierarchy and should be solved with `class extends`.

I just don't want to make this PR more bigger.

File: verilog/CST/tasks_test.cc
Comment (fangism) [reply verilog/CST/tasks_test.cc:245]: The syntax should be `function static ...` or `function automatic ...`

**Review by fangism — APPROVED** [review]

**Comment by hzeller** [comment]
After #415 was merged, there are a few conflicts in some files, possibly because functions were added at the roughly same place in files. Could you resolve these, so that this one can be merged ?

**Review by fangism — APPROVED** [review]

File: verilog/CST/functions.h
Comment (fangism) [inline verilog/CST/functions.h:124]: Sorry, just noticed.  Can you add unit-tests for this function?

**Review by fangism — CHANGES_REQUESTED** [review]

File: verilog/CST/functions.h
Comment (fangism) [reply verilog/CST/functions.h:124]: We'll merge this now, if you can start a PR to add the missing tests.

**Review by fangism — APPROVED** [review]

File: verilog/CST/functions.h
Comment (MinaToma) [reply verilog/CST/functions.h:124]: @fangism  I added the tests. Is this what you asked for?

**Review by fangism — APPROVED** [review]

**Comment by MinaToma** [comment]
@hzeller can you take this one from here?

**Comment by hzeller** [comment]
Alright, will do some merging

**Comment by MinaToma** [comment]
moved to PR #429 because of conflicts.

**Pull request closed without merging**
